lib/rubygems/spec_fetcher.rb

```diff
@@ -174,20 +174,35 @@ def suggest_gems_from_name(gem_name, type = :latest, num_results = 5)
     max             = gem_name.size / 2
     names           = available_specs(type).first.values.flatten(1)
 
+    min_length = gem_name.length - max
+    max_length = gem_name.length + max
+
     matches = names.map do |n|
+      len = n.name.length
+      # If the length is min_length or shorter, we've done `max` deletions.
+      # If the length is max_length or longer, we've done `max` insertions.
+      # These would both be rejected later, so we skip early for performance.
+      next if len < min_length || len > max_length
+
+      # If the gem doesn't support the current platform, bail early.
       next unless n.match_platform?
-      [n.name, 0] if n.name.downcase.tr("_-", "").include?(gem_name)
-    end.compact
 
-    if matches.length < num_results
-      matches += names.map do |n|
-        next unless n.match_platform?
-        distance = levenshtein_distance gem_name, n.name.downcase.tr("_-", "")
-        next if distance >= max
-        return [n.name] if distance == 0
-        [n.name, distance]
-      end.compact
-    end
+      # The candidate name, normalized the same as gem_name.
+      normalized_name = n.name.downcase.tr("_-", "")
+
+      # If we found an exact match (after stripping underscores and hyphens),
+      # that's our most likely candidate.
+      # Return it immediately, and skip the rest of the loop.
+      return [n.name] if normalized_name == gem_name
+
+      distance = levenshtein_distance gem_name, normalized_name
+
+      # Skip current candidate, if the edit distance is greater than allowed.
+      next if distance >= max
```

Comment:
oh! I just noticed this and will adjust my other comment. We don't even allow `max` distance. That means a 3 character or less gem name gets no suggestions.

I didn't notice this and had assumed that the max was inclusive. This alters the behavior quite a bit.

Ok, given this stringent requirement, we need to simply not run suggestions for a gem name with 3 characters or less since it will never find any gems with a distance <= 1.

Additionally, the length filtering above can go back to `<=` and `>=` and a 4 character gem name is the first input size that gets any suggestions (and it must only be off by 1).

Reply:
Changed back, explicitly filtered out 3-character-or-less gem names, and added a test for that.